File: treeadapter/src/main/java/com/krystianwsul/treeadapter/TreeModelAdapter.java
package com.krystianwsul.treeadapter;

import android.support.annotation.NonNull;
import android.support.v7.widget.RecyclerView;
import android.view.ViewGroup;

public interface TreeModelAdapter {
    @NonNull
    RecyclerView.ViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType);

    boolean hasActionMode();

    void incrementSelected();

    void decrementSelected();
}

File: treeadapter/src/main/java/com/krystianwsul/treeadapter/TreeModelAdapter.kt
package com.krystianwsul.treeadapter

import android.view.ViewGroup

interface TreeModelAdapter<T : TreeHolder> : ActionModeCallback {

    fun onCreateViewHolder(parent: ViewGroup, viewType: Int): T

    fun scrollToTop() = Unit

    fun mutateIds(oldIds: List<Any>, newIds: List<Any>): Pair<List<Any>, List<Any>> = Pair(oldIds, newIds)
}
